tests/python/unittest/test_higher_order_grad.py

```diff
@@ -27,52 +27,79 @@ def test_log():
     def log(x):
         return nd.log(x)
 
+    def grad_op(x):
+        return 1/x
+
     def grad_grad_op(x):
         return -1/(x**2)
 
     arrays = random_arrays((2, 2), (2, 3), (4, 5, 2), (3, 1, 4, 5))
 
     for array in arrays:
-        check_second_order_unary(array, log, grad_grad_op)
+        check_second_order_unary(array, log, grad_op, grad_grad_op)
 
 
 @with_seed()
 def test_log2():
     def log2(x):
         return nd.log2(x)
 
+    def grad_op(x):
+        return 1/(x * math.log(2))
+
     def grad_grad_op(x):
         return -1/((x**2) * math.log(2))
 
     arrays = random_arrays((2, 2), (2, 3), (4, 5, 2), (3, 1, 4, 5))
 
     for array in arrays:
-        check_second_order_unary(array, log2, grad_grad_op)
+        check_second_order_unary(array, log2, grad_op, grad_grad_op)
 
 
 @with_seed()
 def test_log10():
     def log10(x):
         return nd.log10(x)
 
+    def grad_op(x):
+        return 1/(x * math.log(10))
+
     def grad_grad_op(x):
         return -1/((x**2) * math.log(10))
 
     arrays = random_arrays((2, 2), (2, 3), (4, 5, 2), (3, 1, 4, 5))
 
     for array in arrays:
-        check_second_order_unary(array, log10, grad_grad_op)
+        check_second_order_unary(array, log10, grad_op, grad_grad_op)
 
 
-def check_second_order_unary(x, op, grad_grad_op):
+def check_second_order_unary(x, op, grad_op, grad_grad_op):
     x = nd.array(x)
-    expect_grad_grad = grad_grad_op(x)
+    grad_x = grad_op(x)
+    grad_grad_x = grad_grad_op(x)
     x.attach_grad()
+
+    # Manual head_grads.
+    head_grads = nd.random.normal(shape=x.shape)
+    head_grad_grads = nd.random.normal(shape=x.shape)
+    head_grads.attach_grad()
+
+    # Perform compute.
     with autograd.record():
         y = op(x)
-        y_grad = autograd.grad(y, x, create_graph=True, retain_graph=True)[0]
-    y_grad.backward()
-    assert_almost_equal(expect_grad_grad.asnumpy(), x.grad.asnumpy())
+        y_grad = autograd.grad(y, x, head_grads=head_grads,
+                               create_graph=True, retain_graph=True)[0]
+
+    y_grad.backward(head_grad_grads)
+
+    # Compute expected values.
+    expected_grad_grad = grad_grad_x.asnumpy() * head_grad_grads.asnumpy() * \
+        head_grads.asnumpy()
+    expected_heads_grad = grad_x.asnumpy()
+
+    # Validate the gradients.
+    assert_almost_equal(expected_grad_grad, x.grad.asnumpy())
+    assert_almost_equal(expected_heads_grad, head_grads.grad.asnumpy())
```

Comment:
Now I understand your question, i don't think anything is updating head_grads here. Why do you want to set the head gradients manually? To verify your fix?

Reply:
can you try 

`y_grad_grad = autograd.grad(y_grad, x, ..., create_graph = False...)[0]`


and in validation
`assert_almost_equal(expected_heads_grad, y_grad_grad.asnumpy())`

Reply:
Yeah to verify the fix. 
I expected `y_grad.backward(head_grad_grads)` to update the `head_grads.grad` similar to the Pytorch Script from the description.

Thanks for the suggestion,
I will surely try that.

Reply:
y_grad.backward(head_grad_grads) indicate that head_grad_grads are the head gradients, it will update all the input variables which have attached gradient, in this case head_grad_grads is not an input to the graph, so your problem that the grad doesn't get updated is expected:

https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/ndarray/ndarray.py#L2188
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/autograd.py#L270

Reply:
We are checking gradients for `head_grads` (not `head_grad_grads`), which is used to compute `x_grad`, so I believe we should accumulate some gradient in `head_grads`

Reply:
the default behaviour in python vs mxnet is different with respect accumulation of gradients (pytorch: add) mxnet: write. Having said that, I still don't understand why do you expect gradient accumulation in head_grads.

Reply:
Oh. I just expected it to have gradients ( by accumulation or writing ), as it is / its value is used while computing the `x_grad`. But from your and @apeforest's explanation, I kinda understand the behaviour better. 
Thank You for digging in and explaining.